---
title: "Automatic cell-type annotation"
date: '`r Sys.Date()`'
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Automatic cell-type annotation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = FALSE,
  message = FALSE,
  # error = FALSE,
  warning = FALSE,
  eval = FALSE,
  fig.align = "center"
)
options(width = 80)
```

**scPlant** have implemented several different solutions for automatic cell-type annotation that are based on popular tools, including [SingleR](https://doi.org/10.1038/s41590-018-0276-y), [Garnett](https://doi.org/10.1038/s41592-019-0535-3), [scCATCH](https://doi.org/10.1016/j.isci.2020.100882), [Celaref](https://www.bioconductor.org/packages/release/bioc/html/celaref.html) and [CellAssign](https://doi.org/10.1038/s41592-019-0529-1).

```{r}
library(scPlant)
```

## Reference based tools

`SingleR`, `Celaref` and `CellFunTopic` need reference data to perform automatic cell-type annotation.

Please see the following examples. Take `seuratObj_ref` as the reference scRNA-seq data, and take `SeuratObj` as the query un-annotated scRNA-seq data.

#### Using SingleR

```{r}
seuratObj_ref$label <- seuratObj_ref$cell_type
SCE_ref <- Seurat::as.SingleCellExperiment(seuratObj_ref)
SeuratObj <- AutoAnnotate_SingleR(SeuratObj, SCE_ref, ref_type = "single-cell")
# The predicted cell type label is stored in SeuratObj$predicted_label.
```

#### Using Celaref

```{r}
cellinfo_ref <- seuratObj_ref$cell_type %>% tibble::enframe(name = "CellId", value = "Cluster") %>% as.data.frame
result <- AutoAnnotate_Celaref(SeuratObj, counts_ref = seuratObj_ref@assays$RNA@data, 
                               cellinfo_ref = cellinfo_ref, plot = TRUE)
```

#### Using CellFunTopic

We suggest performing GSEA analysis on the reference dataset before transferring annotation to query dataset using CellFunTopic.

```{r}
seuratObj_ref <- RunGSEA_plant(seuratObj_ref, by = 'GO', GeneIDtype = 'TAIR')
result <- CellFunTopic::predictFun(
  query_SeuratObj =  SeuratObj,
  seuratObj_ref,
  group_by = "cell_type",
  cluster_by = "seurat_clusters",
  species = "Arabidopsis thaliana"
)
```


## Marker based tools

`scCATCH`, `Garnett` and `CellAssign` need marker genes to perform automatic cell-type annotation.

To run the following codes, a data frame with two columns (`celltype`, `gene`) is needed, containing the markers of each cell type. It is named `markerDF` here.


#### Using scCATCH

First, we need to build a data frame of marker genes with the required format.

```{r}
marker_custom <- get_marker_scCATCH(markerDF, species = 'Arabidopsis thaliana', tissue = 'root')
```

Then, you can perform automatic annotation with scCATCH.

```{r}
result <- AutoAnnotate_scCATCH(SeuratObj, marker_custom)
```

#### Using Garnett

You can build a marker file by referring to the [document](https://cole-trapnell-lab.github.io/garnett/docs/), or just run the following code. 

```{r}
get_marker_file(markerDF, SeuratObj, file = "markers.txt")
# A marker file named markers.txt will be generated in your working directory.
```

Then, you can perform automatic annotation with Garnett.

```{r}
SeuratObj <- AutoAnnotate_Garnett(SeuratObj, marker_file_path = "markers.txt")
# The predicted cell type label is stored in SeuratObj$predicted_label.
```

You may run into some errors prompting format problem of the marker file while running Garnett. If so, we recommend checking the format of the marker file using [check_markers function](https://cole-trapnell-lab.github.io/garnett/docs/) and modify the format beforehand, which will reduce useless effort and save time.


#### Using CellAssign

```{r}
SeuratObj <- AutoAnnotate_CellAssign(SeuratObj, markerDF)
# The predicted cell type label is stored in SeuratObj$predicted_label.
```













