<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scPlant">
<title>Automatic cell-type annotation â€¢ scPlant</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Automatic cell-type annotation">
<meta property="og:description" content="scPlant">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scPlant</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/scPlant.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/pre-processing.html">Pre-processing</a>
    <a class="dropdown-item" href="../articles/integration.html">Cross-species integration</a>
    <a class="dropdown-item" href="../articles/functional_annotation.html">Functional annotation</a>
    <a class="dropdown-item" href="../articles/regulatory_network.html">Gene regulatory network construction</a>
    <a class="dropdown-item" href="../articles/paired_motif_enrichment.html">Paired Motif Enrichment</a>
    <a class="dropdown-item" href="../articles/auto-annotation.html">Automatic cell-type annotation</a>
    <a class="dropdown-item" href="../articles/deconvolution.html">Cell type decomposition of bulk samples</a>
    <a class="dropdown-item" href="../articles/pseudotime.html">Pseudotime trajectory inference</a>
    <a class="dropdown-item" href="../articles/shiny_app.html">Explore in Built-in Shiny APP</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/compbioNJU/scPlant/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Automatic cell-type annotation</h1>
            
            <h4 data-toc-skip class="date">2023-06-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/compbioNJU/scPlant/blob/HEAD/vignettes/auto-annotation.Rmd" class="external-link"><code>vignettes/auto-annotation.Rmd</code></a></small>
      <div class="d-none name"><code>auto-annotation.Rmd</code></div>
    </div>

    
    
<p><strong>scPlant</strong> have implemented several different solutions
for automatic cell-type annotation that are based on popular tools,
including <a href="https://doi.org/10.1038/s41590-018-0276-y" class="external-link">SingleR</a>, <a href="https://doi.org/10.1038/s41592-019-0535-3" class="external-link">Garnett</a>, <a href="https://doi.org/10.1016/j.isci.2020.100882" class="external-link">scCATCH</a>, <a href="https://www.bioconductor.org/packages/release/bioc/html/celaref.html" class="external-link">Celaref</a>
and <a href="https://doi.org/10.1038/s41592-019-0529-1" class="external-link">CellAssign</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/compbioNJU/scPlant" class="external-link">scPlant</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="reference-based-tools">Reference based tools<a class="anchor" aria-label="anchor" href="#reference-based-tools"></a>
</h2>
<p><code>SingleR</code>, <code>Celaref</code> and
<code>CellFunTopic</code> need reference data to perform automatic
cell-type annotation.</p>
<p>Please see the following examples. Take <code>seuratObj_ref</code> as
the reference scRNA-seq data, and take <code>SeuratObj</code> as the
query un-annotated scRNA-seq data.</p>
<div class="section level4">
<h4 id="using-singler">Using SingleR<a class="anchor" aria-label="anchor" href="#using-singler"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuratObj_ref</span><span class="op">$</span><span class="va">label</span> <span class="op">&lt;-</span> <span class="va">seuratObj_ref</span><span class="op">$</span><span class="va">cell_type</span></span>
<span><span class="va">SCE_ref</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">seuratObj_ref</span><span class="op">)</span></span>
<span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AutoAnnotate_SingleR.html">AutoAnnotate_SingleR</a></span><span class="op">(</span><span class="va">SeuratObj</span>, <span class="va">SCE_ref</span>, ref_type <span class="op">=</span> <span class="st">"single-cell"</span><span class="op">)</span></span>
<span><span class="co"># The predicted cell type label is stored in SeuratObj$predicted_label.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="using-celaref">Using Celaref<a class="anchor" aria-label="anchor" href="#using-celaref"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellinfo_ref</span> <span class="op">&lt;-</span> <span class="va">seuratObj_ref</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">%&gt;%</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/enframe.html" class="external-link">enframe</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"CellId"</span>, value <span class="op">=</span> <span class="st">"Cluster"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">as.data.frame</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AutoAnnotate_Celaref.html">AutoAnnotate_Celaref</a></span><span class="op">(</span><span class="va">SeuratObj</span>, counts_ref <span class="op">=</span> <span class="va">seuratObj_ref</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span>, </span>
<span>                               cellinfo_ref <span class="op">=</span> <span class="va">cellinfo_ref</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="using-cellfuntopic">Using CellFunTopic<a class="anchor" aria-label="anchor" href="#using-cellfuntopic"></a>
</h4>
<p>We suggest performing GSEA analysis on the reference dataset before
transferring annotation to query dataset using CellFunTopic.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuratObj_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunGSEA_plant.html">RunGSEA_plant</a></span><span class="op">(</span><span class="va">seuratObj_ref</span>, by <span class="op">=</span> <span class="st">'GO'</span>, GeneIDtype <span class="op">=</span> <span class="st">'TAIR'</span><span class="op">)</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">CellFunTopic</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/CellFunTopic/man/predictFun.html" class="external-link">predictFun</a></span><span class="op">(</span></span>
<span>  query_SeuratObj <span class="op">=</span>  <span class="va">SeuratObj</span>,</span>
<span>  <span class="va">seuratObj_ref</span>,</span>
<span>  group_by <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>  cluster_by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Arabidopsis thaliana"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="marker-based-tools">Marker based tools<a class="anchor" aria-label="anchor" href="#marker-based-tools"></a>
</h2>
<p><code>scCATCH</code>, <code>Garnett</code> and
<code>CellAssign</code> need marker genes to perform automatic cell-type
annotation.</p>
<div class="section level4">
<h4 id="using-sccatch">Using scCATCH<a class="anchor" aria-label="anchor" href="#using-sccatch"></a>
</h4>
<p>First, we need to build a data frame of marker genes with the
required format.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># marker_custom is a data frame with two columns (celltype, gene), containing the markers of each cell type.</span></span>
<span><span class="co"># Specify the species and tissue as you want</span></span>
<span><span class="va">marker_custom</span><span class="op">$</span><span class="va">species</span> <span class="op">&lt;-</span> <span class="st">'Arabidopsis thaliana'</span></span>
<span><span class="va">marker_custom</span><span class="op">$</span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="st">'root'</span></span>
<span><span class="va">marker_custom</span> <span class="op">&lt;-</span> <span class="va">marker_custom</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'species'</span>, <span class="st">'tissue'</span>, <span class="st">'celltype'</span>, <span class="st">'gene'</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">marker_custom</span><span class="op">$</span><span class="va">pmid</span> <span class="op">&lt;-</span> <span class="fl">11111111</span></span>
<span><span class="va">marker_custom</span><span class="op">$</span><span class="va">subtype1</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">marker_custom</span><span class="op">$</span><span class="va">subtype2</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">marker_custom</span><span class="op">$</span><span class="va">subtype3</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span></code></pre></div>
<p>Then, you can perform automatic annotation with scCATCH.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AutoAnnotate_scCATCH.html">AutoAnnotate_scCATCH</a></span><span class="op">(</span><span class="va">SeuratObj</span>, <span class="va">marker_custom</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="using-garnett">Using Garnett<a class="anchor" aria-label="anchor" href="#using-garnett"></a>
</h4>
<p>You can build a marker file by referring to the <a href="https://cole-trapnell-lab.github.io/garnett/docs/" class="external-link">document</a>,
or just run the following code.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># markerDF is a data frame with two columns (celltype, gene), containing the markers of each cell type.</span></span>
<span><span class="va">markerDF</span> <span class="op">&lt;-</span> <span class="va">markerDF</span><span class="op">[</span><span class="va">markerDF</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">SeuratObj</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">marker_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">markerDF</span><span class="op">$</span><span class="va">gene</span>, <span class="va">markerDF</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sink.html" class="external-link">sink</a></span><span class="op">(</span><span class="st">"markers.txt"</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">x</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">marker_list</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"&gt;"</span>, <span class="va">x</span>, <span class="st">"\n"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"expressed: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">marker_list</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span>, <span class="st">"\n"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sink.html" class="external-link">sink</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Then, you can perform automatic annotation with Garnett.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AutoAnnotate_Garnett.html">AutoAnnotate_Garnett</a></span><span class="op">(</span><span class="va">SeuratObj</span>, marker_file_path <span class="op">=</span> <span class="st">"markers.txt"</span><span class="op">)</span></span>
<span><span class="co"># The predicted cell type label is stored in SeuratObj$predicted_label.</span></span></code></pre></div>
<p>You may run into some errors prompting format problem of the marker
file while running Garnett. If so, we recommend checking the format of
the marker file using <a href="https://cole-trapnell-lab.github.io/garnett/docs/" class="external-link">check_markers
function</a> and modify the format beforehand, which will reduce useless
effort and save time.</p>
</div>
<div class="section level4">
<h4 id="using-cellassign">Using CellAssign<a class="anchor" aria-label="anchor" href="#using-cellassign"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># markerDF is a data frame with two columns (celltype, gene), containing the markers of each cell type.</span></span>
<span><span class="va">marker_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">markerDF</span><span class="op">$</span><span class="va">gene</span>, <span class="va">markerDF</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AutoAnnotate_CellAssign.html">AutoAnnotate_CellAssign</a></span><span class="op">(</span><span class="va">SeuratObj</span>, <span class="va">marker_list</span><span class="op">)</span></span>
<span><span class="co"># The predicted cell type label is stored in SeuratObj$predicted_label.</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Shanni Cao.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
