<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scPlant">
<title>Paired Motif Enrichment • scPlant</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Paired Motif Enrichment">
<meta property="og:description" content="scPlant">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scPlant</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/scPlant.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/auto-annotation.html">Automatic cell-type annotation</a>
    <a class="dropdown-item" href="../articles/integration.html">Cross-species integration</a>
    <a class="dropdown-item" href="../articles/paired_motif_enrichment.html">Paired Motif Enrichment</a>
    <a class="dropdown-item" href="../articles/pre-processing.html">Pre-processing</a>
    <a class="dropdown-item" href="../articles/regulatory_network.html">Gene regulatory network construction</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/compbioNJU/scPlant/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Paired Motif Enrichment</h1>
            
            <h4 data-toc-skip class="date">2023-01-06</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/compbioNJU/scPlant/blob/HEAD/vignettes/paired_motif_enrichment.Rmd" class="external-link"><code>vignettes/paired_motif_enrichment.Rmd</code></a></small>
      <div class="d-none name"><code>paired_motif_enrichment.Rmd</code></div>
    </div>

    
    
<p><strong>scPlant</strong> pipeline have adapted the Paired Motif
Enrichment Tool (PMET) (<a href="https://academic.oup.com/plcell/article/32/9/2742/6118120" class="external-link">Rich-Griffin
et al., 2020</a>) to predict pairs of TF binding motifs within the
promoter regions of cell-type specific marker genes. Three plant species
(<em>Arabidopsis thaliana</em>, <em>Oryza sativa</em>, <em>Zea
mays</em>) are supported for now, and more plant species will be
supported soon.</p>
<div class="section level2">
<h2 id="download-and-installation">Download and installation<a class="anchor" aria-label="anchor" href="#download-and-installation"></a>
</h2>
<div class="section level3">
<h3 id="download-databases-and-scripts">Download databases and scripts<a class="anchor" aria-label="anchor" href="#download-databases-and-scripts"></a>
</h3>
<p>The required databases and scripts to perform Paired Motif Enrichment
analysis can be downloaded <a href="https://biobigdata.nju.edu.cn/scPlantDB/scPlantData.zip" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download in linux system:</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://biobigdata.nju.edu.cn/scPlantDB/scPlantData.zip</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span> scPlantData.zip</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> scPlantData</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="download-genome-files">Download genome files<a class="anchor" aria-label="anchor" href="#download-genome-files"></a>
</h3>
<p>Please download the genome and gff3 files suitable for your scRNA-seq
data. If you already have these files, please copy them to the path
<code>scPlantData/PMET/</code>.</p>
<p>For example, download Arabidopsis thaliana’s <strong>TAIR10</strong>
<a href="https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-55/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz" class="external-link">genome</a>
and <a href="https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-55/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.55.gff3.gz" class="external-link">gff3</a>:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># To make the program works, we recommend downloading the files to the directory scPlantData/PMET/</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> PMET/arabidopsis_thaliana.fa.gz <span class="st">'https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-55/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz'</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> PMET/arabidopsis_thaliana.gff3.gz <span class="st">'https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-55/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.55.gff3.gz'</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> PMET/arabidopsis_thaliana.fa.gz</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> PMET/arabidopsis_thaliana.gff3.gz</span></code></pre></div>
<details><summary><strong>Download other species’s genome and gff3 files</strong>
</summary><p>Users can download other plant organism’s genome and gff3 files in <a href="https://plants.ensembl.org/index.html" class="external-link">EnsemblPlants</a> and <a href="https://phytozome-next.jgi.doe.gov/" class="external-link">Phytozome</a>. For example,
download Zea mays’s <strong>Zm-B73-REFERENCE-NAM-5.0</strong> <a href="https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-55/fasta/zea_mays/dna/Zea_mays.Zm-B73-REFERENCE-NAM-5.0.dna.toplevel.fa.gz" class="external-link">genome</a>
and <a href="https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-55/gff3/zea_mays/Zea_mays.Zm-B73-REFERENCE-NAM-5.0.55.gff3.gz" class="external-link">gff3</a>:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> PMET/zea_mays.fa.gz <span class="st">'https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-55/fasta/zea_mays/dna/Zea_mays.Zm-B73-REFERENCE-NAM-5.0.dna.toplevel.fa.gz'</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> PMET/zea_mays.gff3.gz <span class="st">'https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-55/gff3/zea_mays/Zea_mays.Zm-B73-REFERENCE-NAM-5.0.55.gff3.gz'</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> PMET/zea_mays.fa.gz</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> PMET/zea_mays.gff3.gz</span></code></pre></div>
</details>
</div>
<div class="section level3">
<h3 id="install-dependent-softwares">Install dependent softwares<a class="anchor" aria-label="anchor" href="#install-dependent-softwares"></a>
</h3>
<p>Please make sure several dependent softwares have been installed,
including <code>samtools</code>, <code>bedtools</code>,
<code>meme</code>, <code>parallel</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="paired-motif-enrichment-analysis">Paired Motif Enrichment analysis<a class="anchor" aria-label="anchor" href="#paired-motif-enrichment-analysis"></a>
</h2>
<div class="section level4">
<h4 id="help-information">Help information:<a class="anchor" aria-label="anchor" href="#help-information"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> RunPMET.sh <span class="at">-h</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">USAGE:</span> RunPMET [options]</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Perform</span> Paired Motif Enrichment Test.</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="ex">-r</span> <span class="op">&lt;</span>PMET_path<span class="op">&gt;</span>  : Full path of directory where PMET scripts exist. Required.</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="ex">-o</span> <span class="op">&lt;</span>output_directory<span class="op">&gt;</span> : Output directory for results.</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="ex">-m</span> <span class="op">&lt;</span>gene_inputs<span class="op">&gt;</span>        : File containing genes to be tested. Required. A 2 column tab-delimited text file containing cluster ID in column 1 and genes in column 2.</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="ex">-s</span> <span class="op">&lt;</span>species<span class="op">&gt;</span>  : Species. currently At <span class="er">(</span><span class="ex">Arabidopsis</span> thaliana<span class="kw">)</span><span class="ex">,</span> Os <span class="er">(</span><span class="ex">Oryza</span> sativa<span class="kw">)</span><span class="ex">,</span> Zm <span class="er">(</span><span class="ex">Zea</span> mays<span class="kw">)</span> <span class="ex">are</span> supported.</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="prepare-gene-input-file">Prepare gene input file<a class="anchor" aria-label="anchor" href="#prepare-gene-input-file"></a>
</h4>
<p>Take a true scRNA-seq data of Arabidopsis thaliana (<a href="https://www.sciencedirect.com/science/article/pii/S1674205219301339" class="external-link">Zhang
et al., 2019</a>) as example data (download <a href="https://biobigdata.nju.edu.cn/scPlantDB/SeuratObj.rds" class="external-link">here</a>),
you can use the R codes below to prepare gene input file:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">DEG</span> <span class="op">&lt;-</span> <span class="va">SeuratObj</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">topMarker</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">(</span><span class="va">pct.1</span><span class="op">&gt;</span><span class="va">pct.2</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">p_val</span> <span class="op">&lt;</span><span class="fl">0.05</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">pct.1</span><span class="op">&gt;</span><span class="fl">0.25</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">avg_logFC</span><span class="op">&gt;</span><span class="fl">0.25</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cluster</span>, <span class="va">gene</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">DEG</span>, file <span class="op">=</span> <span class="st">"DEG.txt"</span>, quote <span class="op">=</span> <span class="cn">F</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, row.names <span class="op">=</span> <span class="cn">F</span>, col.names <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="run-the-script">Run the script<a class="anchor" aria-label="anchor" href="#run-the-script"></a>
</h4>
<p>Note: Running the shell script could be time-consuming, particularly
step 2.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> RunPMET.sh <span class="at">-r</span> ./PMET <span class="at">-o</span> output <span class="at">-m</span> DEG.txt <span class="at">-s</span> At</span></code></pre></div>
<p>Once the program has run successfully, you can find result files in
the output directory <code>output/output.txt</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="process-the-result-and-visualization">Process the result and visualization<a class="anchor" aria-label="anchor" href="#process-the-result-and-visualization"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/compbioNJU/scPlant" class="external-link">scPlant</a></span><span class="op">)</span></span></code></pre></div>
<p>Pre-process original PMET output：</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PMETresult</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/processPMET.html">processPMET</a></span><span class="op">(</span><span class="st">"output/output.txt"</span><span class="op">)</span></span></code></pre></div>
<p>Heatmap of PMET result：</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PMEThmp.html">PMEThmp</a></span><span class="op">(</span><span class="va">PMETresult</span>, topn <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="paired_motif_enrichment_files/figure-html/unnamed-chunk-11-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Draw triangle heatmap of PMET result:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/triHmp.html">triHmp</a></span><span class="op">(</span><span class="va">PMETresult</span>, clus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">PMETresult</span><span class="op">$</span><span class="va">Module</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="paired_motif_enrichment_files/figure-html/unnamed-chunk-12-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Network diagram showing top pairs of a cluster:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/topPairsNet.html">topPairsNet</a></span><span class="op">(</span><span class="va">PMETresult</span>, clus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">PMETresult</span><span class="op">$</span><span class="va">Module</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, topn <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="paired_motif_enrichment_files/figure-html/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Shanni Cao.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
