<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scPlant">
<title>Getting Started with scPlant â€¢ scPlant</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started with scPlant">
<meta property="og:description" content="scPlant">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scPlant</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/scPlant.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/pre-processing.html">Pre-processing</a>
    <a class="dropdown-item" href="../articles/integration.html">Cross-species integration</a>
    <a class="dropdown-item" href="../articles/functional_annotation.html">Functional annotation</a>
    <a class="dropdown-item" href="../articles/regulatory_network.html">Gene regulatory network construction</a>
    <a class="dropdown-item" href="../articles/paired_motif_enrichment.html">Paired Motif Enrichment</a>
    <a class="dropdown-item" href="../articles/auto-annotation.html">Automatic cell-type annotation</a>
    <a class="dropdown-item" href="../articles/deconvolution.html">Cell type decomposition of bulk samples</a>
    <a class="dropdown-item" href="../articles/pseudotime.html">Pseudotime trajectory inference</a>
    <a class="dropdown-item" href="../articles/shiny_app.html">Explore in Built-in Shiny APP</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/compbioNJU/scPlant/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting Started with scPlant</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbioNJU/scPlant/blob/HEAD/vignettes/scPlant.Rmd" class="external-link"><code>vignettes/scPlant.Rmd</code></a></small>
      <div class="d-none name"><code>scPlant.Rmd</code></div>
    </div>

    
    
<p>We provide a series of vignettes to help users to get started with
scPlant.</p>
<div class="section level3">
<h3 id="pre-processing">
<a href="pre-processing.html">Pre-processing</a><a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h3>
<p>We recommend <strong>CellFunTopic</strong> package to perform data
pre-processing (quality control, normalization, dimension reduction,
clustering, differential expression analysis, etc.) for convenience.</p>
</div>
<div class="section level3">
<h3 id="cross-species-integration">
<a href="integration.html">Cross-species integration</a><a class="anchor" aria-label="anchor" href="#cross-species-integration"></a>
</h3>
<p><strong>scPlant</strong> allows cross-species integration of
single-cell data in matched organs/tissues using one-by-one orthologous
genes as anchors.</p>
</div>
<div class="section level3">
<h3 id="functional-annotation">
<a href="functional_annotation.html">Functional annotation</a><a class="anchor" aria-label="anchor" href="#functional-annotation"></a>
</h3>
<p><strong>scPlant</strong> provides methods for Gene Set Enrichment
Analysis (GSEA) in model plants. We recommend
<strong>CellFunTopic</strong> package to visualize the GSEA result and
perform topic modelling, revealing cellular programs shared across cell
types or exclusive to a particular cell type.</p>
</div>
<div class="section level3">
<h3 id="gene-regulatory-network-construction">
<a href="regulatory_network.html">Gene regulatory network
construction</a><a class="anchor" aria-label="anchor" href="#gene-regulatory-network-construction"></a>
</h3>
<p><strong>scPlant</strong> pipeline have incorporated the widely used
single-cell GRN analysis tool SCENIC (<a href="https://www.nature.com/articles/nmeth.4463" class="external-link">Aibar et al.,
2017</a>), and prepared required cisTarget databases to apply SCENIC in
plants such as Arabidopsis thaliana, Oryza sativa, Zea mays. A recently
published plant specific GRN inference tool MINI-EX (<a href="https://doi.org/10.1016/j.molp.2022.10.016" class="external-link">Ferrari et al.,
2022</a>) has also been incorporated into the <strong>scPlant</strong>
pipeline.</p>
</div>
<div class="section level3">
<h3 id="paired-motif-enrichment">
<a href="paired_motif_enrichment.html">Paired motif
enrichment</a><a class="anchor" aria-label="anchor" href="#paired-motif-enrichment"></a>
</h3>
<p><strong>scPlant</strong> pipeline have adapted the Paired Motif
Enrichment Tool (PMET) (<a href="https://academic.oup.com/plcell/article/32/9/2742/6118120" class="external-link">Rich-Griffin
et al., 2020</a>) to predict pairs of TF binding motifs within the
promoter regions of cell-type specific marker genes.</p>
</div>
<div class="section level3">
<h3 id="automatic-cell-type-annotation">
<a href="auto-annotation.html">Automatic cell-type
annotation</a><a class="anchor" aria-label="anchor" href="#automatic-cell-type-annotation"></a>
</h3>
<p><strong>scPlant</strong> have implemented several different solutions
for automatic cell-type annotation that are based on popular tools,
including <a href="https://doi.org/10.1038/s41590-018-0276-y" class="external-link">SingleR</a>, <a href="https://doi.org/10.1038/s41592-019-0535-3" class="external-link">Garnett</a>, <a href="https://doi.org/10.1016/j.isci.2020.100882" class="external-link">scCATCH</a>, <a href="https://www.bioconductor.org/packages/release/bioc/html/celaref.html" class="external-link">Celaref</a>
and <a href="https://doi.org/10.1038/s41592-019-0529-1" class="external-link">CellAssign</a>.</p>
</div>
<div class="section level3">
<h3 id="estimating-the-cell-type-composition-of-bulk-samples">
<a href="deconvolution.html">Estimating the cell type composition of
bulk samples</a><a class="anchor" aria-label="anchor" href="#estimating-the-cell-type-composition-of-bulk-samples"></a>
</h3>
<p><strong>scPlant</strong> helps estimating the cell type composition
of bulk samples using a single cell reference. Cell-type deconvolution
was implemented based on the <a href="https://www.nature.com/articles/nmeth.3337" class="external-link">CIBERSORT</a>
algorithm.</p>
</div>
<div class="section level3">
<h3 id="pseudotime-trajectory-inference">
<a href="pseudotime.html">Pseudotime trajectory inference</a><a class="anchor" aria-label="anchor" href="#pseudotime-trajectory-inference"></a>
</h3>
<p><strong>scPlant</strong> offers convenient implementation of
pseudotime trajectory inference by incorporating 4 public tools:
Monolce2, Monolce3, Slingshot and CytoTRACE.</p>
</div>
<div class="section level3">
<h3 id="explore-in-built-in-shiny-app">
<a href="shiny_app.html">Explore in Built-in Shiny APP</a><a class="anchor" aria-label="anchor" href="#explore-in-built-in-shiny-app"></a>
</h3>
<p>The gene regulatory network, differentially expressed genes, and
functional annotation can be explored interactively in the built-in
Shiny web application.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Shanni Cao.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
